experimental/web_client/src/lib.rs

```diff
@@ -0,0 +1,163 @@
+}
+
+// TODO(#1867): Add remote attestation support.
+const TEE_MEASUREMENT: &[u8] = br"Test TEE measurement";
+
+/// gRPC-web implementation of a [`UnaryClient`].
+struct GrpcWebClient {
+    uri: String,
+}
+
+impl GrpcWebClient {
+    pub fn create(uri: &str) -> Self {
+        Self {
+            uri: format!("{}/oak.session.unary.v1.UnarySession/Message", uri),
+        }
+    }
+}
+
+// Not marked as send, as the underlying client uses JavaScript APIs for
+// networking, which are not send.
+#[async_trait(? Send)]
+impl UnaryClient for GrpcWebClient {
+    async fn message(&mut self, session_id: SessionId, body: Vec<u8>) -> anyhow::Result<Vec<u8>> {
+        let reply = grpc_web::grpc_web_unary::<UnaryRequest, UnaryResponse>(
+            &self.uri,
+            UnaryRequest {
+                session_id: session_id.to_vec(),
+                body,
+            },
+        )
+        .await?;
+        Ok(reply.body)
+    }
+}
+
+// The underlying [`GenericAttestationClient`] is held in a `wasm_mutex`, which
+// provides a queue of invocations. Ref: https://lib.rs/crates/wasm_mutex
+// TODO(#2908): Support concurrent invocations.
+type WebClientInner = Rc<wasm_mutex::Mutex<GenericAttestationClient<GrpcWebClient>>>;
+
+#[wasm_bindgen]
+pub struct WebClient {
+    inner: WebClientInner,
+}
+
+#[wasm_bindgen]
+impl WebClient {
+    pub async fn new(uri: String) -> Result<WebClient, String> {
+        let grpc_web_client = GrpcWebClient::create(&uri);
+        let inner = GenericAttestationClient::create(
+            grpc_web_client,
+            TEE_MEASUREMENT,
+            Box::new(|server_identity| {
+                if !server_identity.additional_info.is_empty() {
+                    Ok(())
+                } else {
+                    anyhow::bail!("No additional info provided.")
+                }
+            }),
+        )
+        .await
+        .context("Could not create Oak Functions client")
+        .map_err(|error| error.to_string())?;
+        let inner = Rc::new(wasm_mutex::Mutex::new(inner));
+        Ok(WebClient { inner })
+    }
+
+    // Typically this would an async function. Instead it's a sync function
+    // that returns a JS Promise (the JS equivalent of a rust future).
+    // The reason for this is that we cannot access the stack allocated self
+    // inside a Rust future.
+    // Ref: https://github.com/rustwasm/wasm-bindgen/issues/1858
+    pub fn invoke(&self, request: Vec<u8>) -> js_sys::Promise {
+        let inner = self.inner.clone();
+
+        wasm_bindgen_futures::future_to_promise(async move {
+            let response = WebClient::inner_invoke(inner, request)
+                .await
+                .map_err(|error| error.to_string())?;
+
+            // Get the response body, while removing trailing empty space.
+            let body = &response.body[0..response
+                .length
+                .try_into()
+                .map_err(|_| "Could not get response length")?];
```

Comment:
Quite a lot of stuff in the square brackets, perhaps extract the length to a separate var?